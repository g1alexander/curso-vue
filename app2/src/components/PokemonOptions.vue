<script setup lang="ts">
import { inject, type PropType } from "vue";
import type { PokemonsOptions } from "../helpers/interfaces/PokemonOptions";

interface PokemonId {
  pokemonId: { value: number };
  showPokemon: { value: boolean };
}

const props = defineProps({
  pokemons: {
    type: Array as PropType<PokemonsOptions[]>,
    required: true,
  },
});

let pokemon = inject("pokemon") as PokemonId;

const changeId = (id: number): void => {
  pokemon.pokemonId.value = id;
  pokemon.showPokemon.value = true;
};

defineExpose({
  pokemon,
  changeId,
});
</script>

<template>
  <div class="options-container">
    <ul>
      <li
        @click="changeId(pokemon.id)"
        v-for="pokemon in props.pokemons"
        :key="pokemon.id"
      >
        {{ pokemon.name }}
      </li>
    </ul>
  </div>
</template>

<style></style>
