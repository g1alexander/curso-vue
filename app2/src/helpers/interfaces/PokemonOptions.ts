export interface PokemonsOptions {
  name: string;
  id: number;
}
