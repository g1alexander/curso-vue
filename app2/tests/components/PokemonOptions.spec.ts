import { describe, expect, test } from "vitest";
import { shallowMount } from "@vue/test-utils";
import PokemonOptions from "../../src/components/PokemonOptions.vue";
import { mockPokemons } from "../mocks/pokemon.mock";

describe("PokemonOptions component", () => {
  const wrapper = shallowMount(PokemonOptions, {
    props: {
      pokemons: mockPokemons,
    },
    global: {
      provide: {
        pokemon: {
          pokemonId: { value: 0 },
          showPokemon: { value: false },
        },
      },
    },
  });

  test("should to macth snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("debe de mostrar las 4 opciones correctamente", () => {
    const [li1, li2, li3, li4] = wrapper.findAll("li");

    expect(wrapper.findAll("li").length).toBe(4);

    expect(li1.text()).toBe("bulbasaur");
    expect(li2.text()).toBe("ivysaur");
    expect(li3.text()).toBe("venusaur");
    expect(li4.text()).toBe("charmander");
  });

  test("should", () => {
    const [li1, li2, li3, li4] = wrapper.findAll("li");

    li1.trigger("click");
    expect(wrapper.vm.pokemon.pokemonId.value).toBe(1);

    li2.trigger("click");
    expect(wrapper.vm.pokemon.pokemonId.value).toBe(2);

    li3.trigger("click");
    expect(wrapper.vm.pokemon.pokemonId.value).toBe(3);

    li4.trigger("click");
    expect(wrapper.vm.pokemon.pokemonId.value).toBe(4);
  });
});
