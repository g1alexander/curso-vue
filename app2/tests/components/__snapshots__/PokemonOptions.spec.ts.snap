// Vitest Snapshot v1

exports[`PokemonOptions component > should to macth snapshot 1`] = `
"<div class=\\"options-container\\">
  <ul>
    <li>bulbasaur</li>
    <li>ivysaur</li>
    <li>venusaur</li>
    <li>charmander</li>
  </ul>
</div>"
`;
